[RSDK-5864] add controlled moveStraight

JohnN193 opened the pull request:
https://viam.atlassian.net/browse/RSDK-5864

this PR adds a sensorcontrolled MoveStraight option, with controls on heading and position(in lat/long) of the robot sending velocity commands. All controls use the same velocity PID loop under the hood, to avoid managing multiple sets. Heading can use orientation or compass heading, and will use orientation if both are provided. Planning on adding a "no position" move straight option, where only a velocity sensor is provided. Will include in this PR if its straightforward to add. 

tested using wheeled odometry.

randhid commented on components/base/sensorcontrolled/sensorcontrolled.go:
```diff
@@ -209,12 +261,115 @@ func (sb *sensorBase) isPolling() bool {
+	}
+
+	if sb.position == nil || len(sb.conf.ControlParameters) == 0 {
+		sb.logger.CWarn(ctx, "Position not given, using base MoveStraight")
```
```suggestion
		sb.logger.CWarnf(ctx, "Position reporting sensor not available, and no control loop is configured, using base %s MoveStraight", sb.controlledBase.Name().ShortName())
```

randhid commented on components/base/sensorcontrolled/sensorcontrolled.go:
```diff
@@ -209,12 +261,115 @@ func (sb *sensorBase) isPolling() bool {
+}
 
-	return sb.controlledBase.MoveStraight(ctx, distanceMm, mmPerSec, extra)
+// calcSlowDownDist computes the angle at which the spin should begin to slow down.
```
```suggestion
// calcSlowDownDist computes the distance at which the MoveStraigh call should begin to slow down.
```

randhid commented on components/base/sensorcontrolled/sensorcontrolled.go:
```diff
@@ -209,12 +261,115 @@ func (sb *sensorBase) isPolling() bool {
+
+			// exit if the straight takes too long
+			if time.Since(startTime) > timeOut {
+				sb.logger.CWarn(ctx, "exceeded time for Spin call, stopping base")
```
```suggestion
				sb.logger.CWarn(ctx, "exceeded time for MoveStraightCall, stopping base")
```

JohnN193 commented on components/base/sensorcontrolled/sensorcontrolled.go:
```diff
@@ -209,12 +261,115 @@ func (sb *sensorBase) isPolling() bool {
-	sb.stopLoop()
 	ctx, done := sb.opMgr.New(ctx)
 	defer done()
 	sb.setPolling(false)
```
yeah I currently still need it, as I left the setPolling logic in `Spin` in the previous PR

randhid commented on components/base/sensorcontrolled/sensorcontrolled.go:
```diff
@@ -209,12 +271,130 @@ func (sb *sensorBase) isPolling() bool {
+
+// calcPositionError calculates the current error in position.
+// This results in the distance the base needs to travel to reach the goal.
+func calcPositionError(ctx context.Context, distanceMm int, initPos *geo.Point,
```
nice call calculating the positionLeftToTravel, maybe name that function that?

randhid commented on components/base/sensorcontrolled/sensorcontrolled.go:
```diff
@@ -3,10 +3,12 @@ package sensorcontrolled
 
```
Can we move the MoveStraightFunctions into a `movestraight.go` file on this package?

randhid commented on components/base/sensorcontrolled/sensorcontrolled.go:
```diff
@@ -391,7 +391,7 @@ func calcLinVel(errDist, mmPerSec, slowDownDist float64) float64 {
 }
 
-// calcSlowDownDist computes the angle at which the spin should begin to slow down.
+// calcSlowDownDist computes the distance at which the MoveStraigh call should begin to slow down.
```
```suggestion
// calcSlowDownDist computes the distance at which the MoveStraight call should begin to slow down.
```

randhid commented on components/base/sensorcontrolled/sensorcontrolled.go:
```diff
@@ -209,12 +271,130 @@ func (sb *sensorBase) isPolling() bool {
+			if err != nil {
+				return err
+			}
+			sb.logger.Info(angVelDes)
```
Debug level

JohnN193 commented:
![run182 yaw](https://github.com/viamrobotics/rdk/assets/92045055/fe0d9ad7-87b2-49d5-ba51-98072031183c)
the heading has been controlled

randhid commented on components/base/sensorcontrolled/sensorcontrolled.go:
```diff
@@ -209,12 +271,130 @@ func (sb *sensorBase) isPolling() bool {
+			}
+
+			// update velocity controller
+			if err := sb.updateControlConfig(ctx, linVelDes/1000.0, angVelDes); err != nil {
```
What's the unit conversion from `calcPosError` -> control config's SetVel ask?
We're ultimately converting to something SetPower can use right?

JohnN193 replied:
the control loop handles converting things into `SetPower`. I am only calculating velocities at this layer, where `linVelDes` is `mmPerSec` from the request up until I approach the goal, where I slow down the rover. the `1000` is just converting mm to m since the control loop is configured to use the raw velocity messages from a movement sensor

JohnN193 commented on components/base/sensorcontrolled/sensorcontrolled.go:
```diff
@@ -153,10 +157,32 @@ func (sb *sensorBase) Reconfigure(ctx context.Context, deps resource.Dependencie
 		return errNoGoodSensor
 	}
 
+	sb.determineHeadingFunc(ctx)
```
I can technically use this in `Spin`, but saving that change for another day

martha-johnston commented on components/base/sensorcontrolled/movestraight.go:
```diff
@@ -0,0 +1,191 @@
+) error {
+	ctx, done := sb.opMgr.New(ctx)
+	defer done()
+	sb.setPolling(false)
```
is setPolling necessary again? or is there still a plan to remove it in a future PR

JohnN193 replied:
plan to remove in a future PR, since I left the logic in Spin

martha-johnston commented on components/base/sensorcontrolled/movestraight.go:
```diff
@@ -0,0 +1,191 @@
+	}
+	// make sure the control loop is enabled
+	if sb.loop == nil {
+		if err := sb.startControlLoop(); err != nil {
```
I think you also need to also check here that the control Parameters or the control Config exists.

JohnN193 replied:
kk I missed that you can have control parameters and no loop config set, if the velocities are not set on the base. I had a check on `len(sb.conf.ControlParameters)` but I'll switch that to `sb.controlLoopConfig`

JohnN193 replied:
i misread the reconfigure code so my above comment isn't true. I do not use controls if `len(sb.conf.ControlParameters) == 0`

JohnN193 replied:
switched the check to `len(sb.controlLoopConfig.Blocks)`

martha-johnston reviewed (commented):
left a couple small comments, I think we should go through velocity, spin, and movestraight and maybe clean up for consistency. specifically with setPolling, and when we check the length of controlParameters from the congif vs when we check the length of the controlLoopConfig.

Also, does this now block the motor control loops? Like if you have control loops on both your motors, and a control loop on your base and you call MoveStraight on your base which loop(s) are used?

JohnN193 commented:
> Also, does this now block the motor control loops? Like if you have control loops on both your motors, and a control loop on your base and you call MoveStraight on your base which loop(s) are used?

I'll have to confirm I set it up correctly, but when the base loop is configured the SensorControlled MoveStraight should not use the motor loops.

martha-johnston reviewed (approved):
lgtm pending testing sensor controlled base with two motors that have their own control loops to make sure there's no conflicts. I agree that the motors loops should never be started but just want to double check

JohnN193 commented on components/base/sensorcontrolled/movestraight.go:
```diff
@@ -0,0 +1,204 @@
+	for {
+		select {
+		case <-ctx.Done():
+			// do not return context canceled errors, just log them
+			if errors.Is(ctx.Err(), context.Canceled) {
+				sb.logger.Error(ctx.Err())
+				return nil
+			}
```
switched to logging context canceled errors cc @randhid

randhid replied:
so if you're using the API context in your busy loop, you don't know where that is coming from and what is cancelling it. If you switch to a background context spawned from within this code that you have full control lover, that is the preferred way to do things.

We can follow up context clean up as we test with the grid pattern and sine velocities.

JohnN193 commented on components/base/sensorcontrolled/movestraight.go:
```diff
@@ -0,0 +1,204 @@
+	}
+
+	// the currDist will always return as positive, so we need the goal distanceMm to be positive
+	currDist := initPos.GreatCircleDistance(pos) * 1000000.
+	return math.Abs(float64(distanceMm)) - currDist, nil
```
the distance from the `GreatCircleDistance` calculation always be positive, so I needed the setpoint for the position term to also be positive. negative distances get handled in a different spot

martha-johnston replied:
(maybe it will become more obvious as I get farther into this review, but) is it possible to handle negative distances all in this function as well just so it's all in the same spot?

JohnN193 commented on components/base/sensorcontrolled/movestraight.go:
```diff
@@ -0,0 +1,204 @@
+	// have the velocity slow down when appoaching the goal. Otherwise use the desired velocity
+	linVel := errDist * mmPerSec / slowDownDist
+	absMmPerSec := math.Abs(mmPerSec)
+	if linVel > 0 && linVel > absMmPerSec {
+		return absMmPerSec
+	}
+	if linVel < 0 && linVel < -absMmPerSec {
+		return -absMmPerSec
+	}
```
this logic ended up being kinda gross. based on the direction of the velocity we need to set the upper bound of the desired velocity

JohnN193 commented on components/base/sensorcontrolled/movestraight.go:
```diff
@@ -0,0 +1,204 @@
+
+// calcSlowDownDist computes the distance at which the MoveStraight call should begin to slow down.
+// This helps to prevent overshoot when reaching the goal and reduces the jerk on the robot when the straight is complete.
+func calcSlowDownDist(distanceMm int) float64 {
+	slowDownDist := float64(distanceMm) * slowDownDistGain
+	if slowDownDist < -maxSlowDownDist {
+		return -maxSlowDownDist
+	}
+	if slowDownDist > maxSlowDownDist {
+		return maxSlowDownDist
+	}
+	return slowDownDist
+}
```
this function is used to define if the robots velocity should be positive or negative from `distanceMm`. Not the biggest fan of using a 3rd variable to determine this, but `slowDownDist` also has functionality for having the robot slow down when we get close to the goal, so I am okay with this.

JohnN193 commented on components/base/sensorcontrolled/movestraight.go:
```diff
@@ -0,0 +1,204 @@
+// calcLinVel computes the desired linear velocity based on how far the base is from reaching the goal.
+func calcLinVel(errDist, mmPerSec, slowDownDist float64) float64 {
+	// have the velocity slow down when appoaching the goal. Otherwise use the desired velocity
+	linVel := errDist * mmPerSec / slowDownDist
```
`slowDownDist` and `mmPerSec` are used to determine if the velocity should be positive or negative

JohnN193 commented:
added logic for negative signs from inputs. Not the biggest fan of where those landed but they work. lmk if theres anything from them I should clean up. Added comments to everywhere that should be relevant.

added some tests for this PR but skipped the MoveStraight specific tests due to the race conditions we have seen in the control loop before.

martha-johnston commented on components/base/sensorcontrolled/movestraight.go:
```diff
@@ -0,0 +1,204 @@
+// calcSlowDownDist computes the distance at which the MoveStraight call should begin to slow down.
+// This helps to prevent overshoot when reaching the goal and reduces the jerk on the robot when the straight is complete.
+func calcSlowDownDist(distanceMm int) float64 {
+	slowDownDist := float64(distanceMm) * slowDownDistGain
+	if slowDownDist < -maxSlowDownDist {
+		return -maxSlowDownDist
+	}
+	if slowDownDist > maxSlowDownDist {
+		return maxSlowDownDist
+	}
+	return slowDownDist
```
```suggestion
	slowDownDist := float64(distanceMm) * slowDownDistGain
	if math.Abs(slowDownDist) > maxSlowDownDist {
		return maxSlowDownDist * sign(distanceMm)
	}
	return slowDownDist
```

martha-johnston commented on components/base/sensorcontrolled/movestraight.go:
```diff
@@ -0,0 +1,204 @@
+	}
+
+	headingErr := initHeading - currHeading
+	headingErrWrapped := headingErr - (math.Floor((headingErr+180.)/(2*180.)))*2*180. // [-180;180)
```
wrap around math always confuses me... do you have a unit test for this? could be helpful if you haven't seen this scenario with hardware testing

JohnN193 replied:
I can add a test for this yeah. I'll validate that the wrapped angle error falls within +-180 degrees for a range of setpoints(initHeadings in this case)

martha-johnston commented on components/base/sensorcontrolled/movestraight.go:
```diff
@@ -0,0 +1,204 @@
+func calcLinVel(errDist, mmPerSec, slowDownDist float64) float64 {
+	// have the velocity slow down when appoaching the goal. Otherwise use the desired velocity
+	linVel := errDist * mmPerSec / slowDownDist
+	absMmPerSec := math.Abs(mmPerSec)
+	if linVel > 0 && linVel > absMmPerSec {
+		return absMmPerSec
+	}
+	if linVel < 0 && linVel < -absMmPerSec {
+		return -absMmPerSec
+	}
+	return linVel
```
```suggestion
	absMmPerSec := math.Abs(mmPerSec)
	if math.Abs(linVel) > absMmPerSec {
		return absMmPerSec * sign(linVel)
	}
	return linVel
```

martha-johnston reviewed (approved):
overall lgtm, I just have a couple questions:
1. have you tested/measured how accurate the distance control is? maybe even compared to wheeled_base `MoveStraight`?
2. what happens if distanceMm and mmPerSec are both negative? Have you tested this scenario? It seems like you are careful about the signs of everything but you might want to check that case specifically

martha-johnston commented on components/base/sensorcontrolled/movestraight.go:
```diff
@@ -0,0 +1,198 @@
+func calcSlowDownDist(distanceMm int) float64 {
+	slowDownDist := float64(distanceMm) * slowDownDistGain
+	if math.Abs(slowDownDist) > maxSlowDownDist {
+		return maxSlowDownDist * sign(distanceMm)
```
```suggestion
		return maxSlowDownDist * sign(float64(distanceMm))
```

oops

JohnN193 commented:
1. I did a few benchtop runs to validate that the MoveStraight function was working the way I expected & stopping when I wanted, but I did not do any measured floor tests because of the single encoder issues we were seeing with spin
2. yeah I made sure to cover that scenario, its the reason why the distance error always being positive was so annoying. The `slowDownDist` and `mmPerSec` signs are setup to cancel each other out in that scenario.

martha-johnston commented:
> I did a few benchtop runs to validate that the MoveStraight function was working the way I expected & stopping when I wanted, but I did not do any measured floor tests because of the single encoder issues we were seeing with spin

okay we should test this when we are doing sensor controlled base tests later today. Just curious to see if it's wildly different
